tests/Data/Text/Utf8/BoyerMooreSpec.hs

```diff
@@ -0,0 +1,253 @@
+    -- generator to cover the entire range of code points.
+    genChar = Gen.frequency
+      [ (4, QuickCheck.arbitraryASCIIChar)
+      , (1, QuickCheck.arbitrary)
+      ]
+    genNonEmptyText = do
+      chars <- Gen.listOf1 genChar
+      pure $ Text.pack chars
+
+  fragments <- Gen.listOf1 $ Gen.resize 5 genNonEmptyText
+  let
+    genFragment = Gen.elements $ filter (not . Text.null) fragments
+    genSmall = Gen.scale (`div` 3) $ Gen.listOf1 genFragment
+    genBig = Gen.scale (* 4) $ Gen.listOf1 genFragment
+
+  needle <- fmap Text.concat genSmall
+  haystack <- fmap Text.concat genBig
+  pure (needle, haystack)
+
+spec :: Spec
+spec = parallel $ modifyMaxSuccess (const 200) $ do
+  describe "build" $ do
+    prop "does not throw exceptions" $ \ (pat :: Text) ->
+      rnf $ BoyerMoore.buildAutomaton pat
+
+  describe "runText" $ do
+
+    describe "when given a needle equal to the haystack" $ do
+
+      it "reports a single match for a repeated character" $
+        forM_ [1..128] $ \n ->
+          needleIsHaystackMatches $ Text.replicate n "a"
+
+      it "reports a single match for non-BMP data" $ do
+        -- Include a few code points outside of the Basic Multilingual Plane,
+        -- which require multible code units to encode.
+        needleIsHaystackMatches "\x000437b8suffix"
+        needleIsHaystackMatches "aaa\359339aaa\95759aa\899256aa"
+
+      prop "reports a single match for random needles" $ \needle ->
+        not (Text.null needle) ==> needleIsHaystackMatches needle
+
+    describe "when given a sliced text (with nonzero internal offset)" $ do
+
+      it "still reports offset relative to the text start" $
+        -- The match position should be relative to the start of the text "a".
+        -- Even if this text is represented as a slice of "bbba" internally.
+        matchEndPositions "a" (Text.dropWhile (== 'b') "bbba") `shouldMatchList` [1]
+
+    describe "when given non-ascii inputs" $ do
+
+      -- We have a special lookup table for bad character shifts for
+      -- the first 128 code units, which is always hit for ascii inputs.
+      -- Also exercise the fallback code path with a different input.
+      -- The code point é is encoded as two code units in UTF-8.
+      -- 0             7          13
+      -- │             │           │
+      -- ▼             ▼           ▼
+      -- ┌───┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┐
+      -- │ é │c│l│a│i│r│e│c│l│a│i│r│ Code Points
+      -- ├─┬─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┤
+      -- │ │ │ │ │ │ │ │ │ │ │ │ │ │ Code Units (Bytes)
+      -- └─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┘
+      it "reports a match if the haystack contains a character > U+7f" $ do
+        matchEndPositions "eclair" "éclaireclair" `shouldMatchList` [13]
+        matchEndPositions "éclair" "éclaireclair" `shouldMatchList` [7]
+        matchEndPositions "éclair" "eclairéclair" `shouldMatchList` [13]
+
+      it "reports the correct UTF-16 index for surrogate pairs" $ do
```

Comment:
The test is good, but the description is out of date.

Reply:
Done!